File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/nudg++/trunk/Src/Codes2D/CubatureVolumeMesh2D.cpp
// CubatureVolumeMesh2D.m
// function cub = CubatureVolumeMesh2D(Corder)
// 2007/06/06
//---------------------------------------------------------
#include "NDGLib_headers.h"
#include "NDG2D.h"


//---------------------------------------------------------
Cub2D& NDG2D::CubatureVolumeMesh2D(int Corder)
//---------------------------------------------------------
{
  // function cub = CubatureVolumeMesh2D(Corder)
  // purpose: build cubature nodes, weights and geometric factors for all elements
  //
  // Note: m_cub is member of Globals2D

  // set up cubature nodes
  Cubature2D(Corder, m_cub);

  // evaluate generalized Vandermonde of Lagrange interpolant functions at cubature nodes
  InterpMatrix2D(m_cub);

  // evaluate local derivatives of Lagrange interpolants at cubature nodes
  Dmatrices2D(this->N, m_cub);

  // evaluate the geometric factors at the cubature nodes
  GeometricFactors2D(m_cub);

  // custom mass matrix per element
  DMat mmk; DMat_Diag D; DVec d;
  m_cub.mmCHOL.resize(Np*Np, K);
  m_cub.mm    .resize(Np*Np, K);

  for (int k=1; k<=K; ++k) {
    d=m_cub.J(All,k); d*=m_cub.w; D.diag(d);  // weighted diagonal
    mmk = m_cub.VT * D * m_cub.V;     // mass matrix for element k
    m_cub.mm(All,k)     = mmk;        // store mass matrix
    m_cub.mmCHOL(All,k) = chol(mmk);  // store Cholesky factorization
  }

  // incorporate weights and Jacobian
  m_cub.W = outer(m_cub.w, ones(K));
  m_cub.W.mult_element(m_cub.J);

  // compute coordinates of cubature nodes
  m_cub.x = m_cub.V * this->x;
  m_cub.y = m_cub.V * this->y;

  return m_cub;
}

File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/Codes1.1/Codes2D/CubatureVolumeMesh2D.m
function cub = CubatureVolumeMesh2D(CubatureOrder)

% function cub = CubatureVolumeMesh2D(CubatureOrder)
% purpose: build cubature nodes, weights and geometric factors for all elements

Globals2D;

% set up cubature nodes
[cub.R,cub.S,cub.W, cub.Ncub] = Cubature2D(CubatureOrder); 

% evaluate generalized Vandermonde of Lagrange interpolant functions at cubature nodes
cub.V  = InterpMatrix2D(cub.R, cub.S); 

% evaluate local derivatives of Lagrange interpolants at cubature nodes
[cub.Dr,cub.Ds] = Dmatrices2D(N,cub.R,cub.S,V);

% evaluate the geometric factors at the cubature nodes
[cub.rx,cub.sx,cub.ry,cub.sy,cub.J] = GeometricFactors2D(x,y, cub.Dr,cub.Ds);

% custom mass matrix per element
cub.mmCHOL = zeros(Np, Np, K); cub.mm = zeros(Np, Np, K);
for k=1:K
  cub.mm(:,:,k)     = cub.V'*diag(cub.J(:,k).*cub.W(:))*cub.V;
  cub.mmCHOL(:,:,k) = chol(cub.mm(:,:,k));
end

% incorporate weights and Jacobian
cub.w = cub.W; cub.W = cub.W*ones(1,K); cub.W = cub.W.*cub.J; 

% compute coordinates of cubature nodes
cub.x = cub.V*x; cub.y = cub.V*y;
return
